test/extensions/filters/http/original_src/original_src_test.cc

```diff
@@ -0,0 +1,217 @@
+TEST_F(OriginalSrcHttpTest, DecodeHeadersIpv4AddressBleachesPort) {
+  auto filter = makeDefaultFilter();
+  Network::Socket::OptionsSharedPtr options;
+  setAddressToReturn("tcp://1.2.3.4:80");
+  EXPECT_CALL(callbacks_, addUpstreamSocketOptions(_)).WillOnce(SaveArg<0>(&options));
+
+  filter->decodeHeaders(headers_, false);
+
+  NiceMock<Network::MockConnectionSocket> socket;
+  const auto expected_address = Network::Utility::parseInternetAddress("1.2.3.4");
+  EXPECT_CALL(socket, setLocalAddress(PointeesEq(expected_address)));
+
+  // Not ideal -- we're assuming that the original_src option is first, but it's a fair assumption
+  // for now.
+  options->at(0)->setOption(socket, envoy::api::v2::core::SocketOption::STATE_PREBIND);
+}
+
+TEST_F(OriginalSrcHttpTest, FilterAddsTransparentOption) {
+  if (!ENVOY_SOCKET_IP_TRANSPARENT.has_value()) {
+    // The option isn't supported on this platform. Just skip the test.
+    return;
+  }
+
+  auto filter = makeDefaultFilter();
+  Network::Socket::OptionsSharedPtr options;
+  setAddressToReturn("tcp://1.2.3.4:80");
+  EXPECT_CALL(callbacks_, addUpstreamSocketOptions(_)).WillOnce(SaveArg<0>(&options));
+
+  filter->decodeHeaders(headers_, false);
+
+  auto transparent_option = findOptionDetails(*options, ENVOY_SOCKET_IP_TRANSPARENT,
+                                              envoy::api::v2::core::SocketOption::STATE_PREBIND);
+
+  EXPECT_TRUE(transparent_option.has_value());
+}
+
+TEST_F(OriginalSrcHttpTest, FilterAddsMarkOption) {
+  if (!ENVOY_SOCKET_SO_MARK.has_value()) {
+    // The option isn't supported on this platform. Just skip the test.
+    return;
+  }
+
+  auto filter = makeMarkingFilter(1234);
+  Network::Socket::OptionsSharedPtr options;
+  setAddressToReturn("tcp://1.2.3.4:80");
+  EXPECT_CALL(callbacks_, addUpstreamSocketOptions(_)).WillOnce(SaveArg<0>(&options));
+
+  filter->decodeHeaders(headers_, false);
+
+  auto mark_option = findOptionDetails(*options, ENVOY_SOCKET_SO_MARK,
+                                       envoy::api::v2::core::SocketOption::STATE_PREBIND);
+
+  ASSERT_TRUE(mark_option.has_value());
+  uint32_t value = 1234;
+  absl::string_view value_as_bstr(reinterpret_cast<const char*>(&value), sizeof(value));
+  EXPECT_EQ(value_as_bstr, mark_option->value_);
+}
+
+TEST_F(OriginalSrcHttpTest, Mark0NotAdded) {
+  if (!ENVOY_SOCKET_SO_MARK.has_value()) {
+    // The option isn't supported on this platform. Just skip the test.
+    return;
+  }
+
+  auto filter = makeMarkingFilter(0);
+  Network::Socket::OptionsSharedPtr options;
+  setAddressToReturn("tcp://1.2.3.4:80");
+  EXPECT_CALL(callbacks_, addUpstreamSocketOptions(_)).WillOnce(SaveArg<0>(&options));
+
+  filter->decodeHeaders(headers_, false);
+
+  auto mark_option = findOptionDetails(*options, ENVOY_SOCKET_SO_MARK,
+                                       envoy::api::v2::core::SocketOption::STATE_PREBIND);
+
+  ASSERT_FALSE(mark_option.has_value());
+}
+
+TEST_F(OriginalSrcHttpTest, DecodeDataDoesNothing) {
```

Comment:
These tests that just verify nothing seem a bit strange to me because they're exercising call orders that will never happen: `decodeData` will never be called before `decodeHeaders`, so normally I see tests run through the expected order of of `decode*` calls. 

It's common to maintain state between the callbacks, e.g. for `decodeHeaders` to decide whether something happens in `decodeData`, which means that calling `decodeData` on its own often doesn't prove much. This test in particular that calls `decodeData(, true)` followed by `decodeData(, false)` especially stands out as a bit strange.

I don't feel super strongly about it in the context of this filter due to its simplicity, just wanted to point it out.

Reply:
I'm actually implicitly asserting on nothing being invoked on callbacks -- it's a strict mock with no expectations set.

Any thoughts on how to make that more obvious? Some way of telling the mock that it should expect no calls at all (without needing to enumerate the disallowed calls)? I agree that it does look odd.

Reply:
What I've done in the past is run through the entire filter lifecycle in a single test: decodeHeaders/decodeData/decodeTrailers, using strict mocks to verify that nothing happens on data/trailers. I don't really feel strongly about this so up to you if you want to change this

Reply:
That's reasonable. I'll do that.